Stats to record how often the ClusterState diff mechanism is used successfully

DaveCTurner opened the pull request:
It's believed that using diffs obsoletes the other mechanism for reusing the
bits of the ClusterState that didn't change between updates, but in fact we
don't know for sure how often the diff mechanism works successfully. The stats
collected here will tell us.

PR opened for comments on approach, and because I'm about to get on a plane so it's either now or much later. I think I'd like to add stats on how often these things are received as well as sent in a similar fashion.

DaveCTurner committed:
Add statistics to record how often the ClusterState diff mechanism is used successfully.

It's believed that using diffs obsoletes the other mechanism for reusing the
bits of the ClusterState that didn't change between updates, but in fact we
don't know for sure how often the diff mechanism works successfully. The stats
collected here will tell us.

DaveCTurner committed:
Picky picky

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateAction.java:
```diff
@@ -90,6 +94,10 @@
     private final IncomingClusterStateListener incomingClusterStateListener;
     private final DiscoverySettings discoverySettings;
 
+    private long fullClusterStateSentCount = 0;
```
these should probably be atomic...

jasontedor replied:
It's sneaky. 😇

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Class encapsulating stats about the PublishClusterStateAction
+ */
+public class PublishClusterStateStats implements Writeable, ToXContentFragment {
```
I think you want ToXContentObject here.

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java:
```diff
@@ -0,0 +1,96 @@
+    private final long clusterStateDiffSentCount;
+    private final long incompatibleClusterStateDiffVersionCount;
+
+    public PublishClusterStateStats(long fullClusterStateSentCount,
```
can you java doc the parameters?

DaveCTurner replied:
Params or getters? Or both?

bleskes reviewed (changes requested):
Code wise, I left some minor comments. Looks good otherwise. 

I have some higher level comments:
1) This is currently implemented on the master side. I think it will have more value on the node side - we can see that if a node is struggling it falls back more to full cluster state publishing etc.
2) We need a unit test.
3) Can you also add a rest test - this will allow use to check we do the right thing in a mixed cluster version (if it's above the right version).
4) think this can go into 6.1.

jasontedor commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java:
```diff
@@ -0,0 +1,96 @@
+        this.incompatibleClusterStateDiffVersionCount = incompatibleClusterStateDiffVersionCount;
+    }
+
+    public PublishClusterStateStats(StreamInput streamInput) throws IOException {
```
We normally name these `in`, it leaves for shorter code and it would be consistent with other places in the codebase.

jasontedor commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java:
```diff
@@ -0,0 +1,96 @@
+        out.writeVLong(incompatibleClusterStateDiffVersionCount);
+    }
+
+    static final class Fields {
```
We've been moving away from these inner fields classes, we don't need them to encapsulate some strings.

jasontedor reviewed (commented):
I agree with @bleskes comments, I left a few more.

DaveCTurner committed:
No need for Fields class

DaveCTurner committed:
Shorten arg name

DaveCTurner committed:
XContent not XContentFragment

DaveCTurner committed:
Make accumulators into AtomicLongs

DaveCTurner committed:
Add javadoc

DaveCTurner committed:
Slated for inclusion in 6.1

DaveCTurner committed:
Add stats for receivers of cluster states too

DaveCTurner commented:
Thanks. I think I've addressed all the comments except for the ones asking for more tests, for which I need a bit more guidance. Not sure I like the inconsistency between `incompatibleClusterStateDiffsSentCount` and `compatibleClusterStateDiffReceivedCount` but they were the cleanest things to measure given the existing code. I could do something more complicated to smooth that out if desired.

DaveCTurner committed:
Nit: Diffs -> Diff

DaveCTurner committed:
Misread comment - should have been ToXContentObject

DaveCTurner committed:
Don't bother collecting the master-side stats at all

DaveCTurner committed:
Stick with v7 for now.

There is more work to be done in the 6.x branch before the version guard can be
reduced like this.

This reverts commit f004119ae655d7203725258a2485de3e6a656d86.

DaveCTurner committed:
Missed one

DaveCTurner committed:
Add assertions that the stats are updated as expected

DaveCTurner committed:
Add testPublishClusterStateStats()

DaveCTurner committed:
Remove assertions from existing test

This reverts commit a9a282813f99902be8ce8dda230ace4b732a4097.

DaveCTurner committed:
Better grammar

DaveCTurner committed:
Add REST assertions

DaveCTurner commented:
@jasontedor @bleskes Could you have another look at this?

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateAction.java:
```diff
@@ -90,6 +98,10 @@
     private final IncomingClusterStateListener incomingClusterStateListener;
     private final DiscoverySettings discoverySettings;
 
+    private AtomicLong fullClusterStateReceivedCount = new AtomicLong();
```
can you please make these final?

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateAction.java:
```diff
@@ -90,6 +98,10 @@
 
+    private AtomicLong fullClusterStateReceivedCount = new AtomicLong();
+    private AtomicLong clusterStateDiffReceivedCount = new AtomicLong();
+    private AtomicLong compatibleClusterStateDiffReceivedCount = new AtomicLong();
```
I personally find having a incompatibleDiff count more handy as it's what we'll be looking for. Maybe just have compatibleDiffCount and incompatibleDiffCount (instead of total)?

DaveCTurner replied:
Ok. How picky do we want to be about concurrency here? I can do this by counting the cluster state diff as incompatible until successfully applied (i.e. no exceptions thrown) and then to decrement the `incompatible` count and increment the `compatible` one, but that does mean you could see incorrect values if you get the stats at the wrong moment.

On second thought the `stats()` method is already a bit racy.

I have various other, more complicated, ideas:
- counting more things, to distinguish in-progress applications from rejected ones.
- catching an `IncompatibleClusterStateVersionException` on the way past and updating the stats in the handler.
- keeping the accumulators as-is and calculating the difference in the `stats()` method (potentially also racy, not sure yet).
- introducing locks around stats reads and updates.

Any particular preferences?

bleskes commented on rest-api-spec/src/main/resources/rest-api-spec/test/nodes.stats/30_discovery.yml:
```diff
@@ -26,4 +28,6 @@
   - is_false: nodes.$master.jvm
   - is_true:  nodes.$master.discovery
+  - is_true:  nodes.$master.discovery.cluster_state_queue
+  - is_true:  nodes.$master.discovery.published_cluster_states_received
```
can we also check the sub elements?

DaveCTurner replied:
So `is_true` fails on the subfields because 0 is falsey. I'm leaning towards saying `gte: 0` as a proxy for "is present and numeric".

bleskes reviewed (changes requested):
Thanks @DaveCTurner . This looks great. I left some very minor feedback

jasontedor commented on core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java:
```diff
@@ -34,25 +36,42 @@
     private final PendingClusterStateStats queueStats;
 
-    public DiscoveryStats(PendingClusterStateStats queueStats) {
+    @Nullable
```
I do not think we need the `@Nullable` annotation on a private field (I question its usefulness in general).

DaveCTurner replied:
Should I remove the one above too?

jasontedor reviewed (changes requested):
I agree with @bleskes comments so I'm marking this as request changes.

DaveCTurner committed:
Make count fields final

DaveCTurner committed:
Count incompatible cluster state diffs rather than all of them

DaveCTurner committed:
Assert that the sub-fields are >= 0

bleskes commented:
Sounds good to me

> On 16 Oct 2017, at 11:25 AM, David Turner <notifications@github.com> wrote:
> 
> @DaveCTurner commented on this pull request.
> 
> In rest-api-spec/src/main/resources/rest-api-spec/test/nodes.stats/30_discovery.yml:
> 
> > @@ -26,4 +28,6 @@
>    - is_false: nodes.$master.name
>    - is_false: nodes.$master.jvm
>    - is_true:  nodes.$master.discovery
> +  - is_true:  nodes.$master.discovery.cluster_state_queue
> +  - is_true:  nodes.$master.discovery.published_cluster_states_received
> So is_true fails on the subfields because 0 is falsey. I'm leaning towards saying gte: 0 as a proxy for "is present and numeric".
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

DaveCTurner committed:
No need for @Nullables

DaveCTurner committed:
Skip these tests until v7.x

DaveCTurner committed:
Line length oops

DaveCTurner committed:
Count incompatible states received in exception handler instead

DaveCTurner commented:
@bleskes changed in b589f36faecc1938e1e2cbefcc932bf49aa043d2 as requested. I think that's everything.

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject("published_cluster_states_received");
```
Nit - strictly speaking these are publishing stats, can we open the object with just published cluster stats (drop received). You can maybe received back in the keys, which can be shorted by dropping the cluster states from the key names - it’s implied from the object they’re in.

bleskes reviewed (approved):
LGTM. If Ci is happy I’m happy.

DaveCTurner committed:
Shorter keys in PublishClusterStateStats#toXContent()

DaveCTurner committed:
Fewer linebreaks

DaveCTurner commented:
@bleskes squashed remaining nits. @jasontedor would you have a look too?

DaveCTurner committed:
Fixed up REST API test

DaveCTurner commented:
@jasontedor is this ok with you?

jasontedor commented on core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java:
```diff
@@ -64,4 +80,5 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
     public PendingClusterStateStats getQueueStats() {
         return queueStats;
     }
+    public PublishClusterStateStats getPublishStats() { return publishStats; }
```
Would you add a newline before this method, and wrap the method definition over multiple lines please?

jasontedor commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java:
```diff
@@ -0,0 +1,88 @@
+        return builder;
+    }
+
+    public long getFullClusterStateReceivedCount() { return fullClusterStateReceivedCount; }
```
This method can be package-private.

jasontedor commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java:
```diff
@@ -0,0 +1,88 @@
+
+    public long getFullClusterStateReceivedCount() { return fullClusterStateReceivedCount; }
+
+    public long getIncompatibleClusterStateDiffReceivedCount() { return incompatibleClusterStateDiffReceivedCount; }
```
This method can be package-private.

jasontedor commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java:
```diff
@@ -0,0 +1,88 @@
+
+    public long getIncompatibleClusterStateDiffReceivedCount() { return incompatibleClusterStateDiffReceivedCount; }
+
+    public long getCompatibleClusterStateDiffReceivedCount() { return compatibleClusterStateDiffReceivedCount; }
```
This method can be package-private.

jasontedor commented on core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java:
```diff
@@ -0,0 +1,88 @@
+        builder.field("incompatible_diffs", incompatibleClusterStateDiffReceivedCount);
+        builder.field("compatible_diffs", compatibleClusterStateDiffReceivedCount);
+        builder.endObject();
+        return builder;
```
Here you can do something like this:

```diff
diff --git a/core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java b/core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java
index 356b9a29dc..36794e880f 100644
--- a/core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java
+++ b/core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateStats.java
@@ -65,9 +65,11 @@ public class PublishClusterStateStats implements Writeable, ToXContentObject {
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject("published_cluster_states");
-        builder.field("full_states", fullClusterStateReceivedCount);
-        builder.field("incompatible_diffs", incompatibleClusterStateDiffReceivedCount);
-        builder.field("compatible_diffs", compatibleClusterStateDiffReceivedCount);
+        {
+            builder.field("full_states", fullClusterStateReceivedCount);
+            builder.field("incompatible_diffs", incompatibleClusterStateDiffReceivedCount);
+            builder.field("compatible_diffs", compatibleClusterStateDiffReceivedCount);
+        }
         builder.endObject();
         return builder;
     }
```

which makes the JSON-structure clearer in the code.

jasontedor commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -413,7 +413,8 @@ public void onNewClusterStateFailed(Exception e) {
         PendingClusterStateStats queueStats = pendingStatesQueue.stats();
-        return new DiscoveryStats(queueStats);
+        PublishClusterStateStats publishStats = publishClusterState.stats();
+        return new DiscoveryStats(queueStats, publishStats);
```
I think:

```diff
diff --git a/core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java b/core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java
index 5f62ee6b97..d688a5d5cd 100644
--- a/core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java
+++ b/core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java
@@ -412,9 +412,7 @@ public class ZenDiscovery extends AbstractLifecycleComponent implements Discover
 
     @Override
     public DiscoveryStats stats() {
-        PendingClusterStateStats queueStats = pendingStatesQueue.stats();
-        PublishClusterStateStats publishStats = publishClusterState.stats();
-        return new DiscoveryStats(queueStats, publishStats);
+        return new DiscoveryStats(pendingStatesQueue.stats(), publishClusterState.stats());
     }
 
     public DiscoverySettings getDiscoverySettings() {
```

is fine?

jasontedor reviewed (commented):
I left a few more comments.

DaveCTurner committed:
Merge branch 'master' into 2017-10-11-discovery-stats

DaveCTurner committed:
Formatting

DaveCTurner committed:
Inline

DaveCTurner committed:
Make getters package-private

DaveCTurner commented:
All addressed @jasontedor - could you review again?

jasontedor reviewed (approved):
LGTM.

DaveCTurner commented:
Thanks both. The backport is on its way.
